src/core/useAtom.ts

```diff
@@ -75,8 +53,33 @@ export function useAtom<Value, Update>(
   }
 
   const ScopeContext = getScopeContext(scope)
-  const [store, mutableSource] = useContext(ScopeContext)
-  const value = useMutableSource(mutableSource, getAtomValue, subscribe)
+  const [store] = useContext(ScopeContext)
+
+  const getAtomValue = useCallback(() => {
+    const atomState = store[READ_ATOM](atom)
+    if (atomState.e) {
+      throw atomState.e // read error
+    }
+    if (atomState.p) {
+      throw atomState.p // read promise
+    }
+    if (atomState.w) {
+      throw atomState.w // write promise
+    }
+    if ('v' in atomState) {
+      return atomState.v as Value
+    }
+    throw new Error('no atom value')
+  }, [store, atom])
+
+  const [value, forceUpdate] = useReducer(getAtomValue, undefined, getAtomValue)
```

Comment:
I never tried using a callback function as the reducer for `useReducer` - What happens when this callback is recreated due to dependency changes - Is it run again or only on dispatch?

Reply:
It doesn't have to be wrapped by `useCallback`. For more info, [this blog post](https://overreacted.io/a-complete-guide-to-useeffect/#why-usereducer-is-the-cheat-mode-of-hooks) might help.
I'm not saying this is ideal, but it's simpler than the previous uMS emulation and more readable. I'm sure we will revisit this in the future.